diff --git a/lib/private/Federation/CloudFederationNotification.php b/lib/private/Federation/CloudFederationNotification.php
index 6ae805df1d98c..d8f6d87f3e237 100644
--- a/lib/private/Federation/CloudFederationNotification.php
+++ b/lib/private/Federation/CloudFederationNotification.php
@@ -32,7 +32,7 @@ public function setMessage($notificationType, $resourceType, $providerId, array
 		$this->message = [
 			'notificationType' => $notificationType,
 			'resourceType' => $resourceType,
-			'providerId' => $providerId,
+			'providerId' => (string)$providerId,
 			'notification' => $notification,
 		];
 	}
diff --git a/lib/private/Federation/CloudFederationShare.php b/lib/private/Federation/CloudFederationShare.php
index 2eb06b3acea05..6bd35cea763e4 100644
--- a/lib/private/Federation/CloudFederationShare.php
+++ b/lib/private/Federation/CloudFederationShare.php
@@ -124,7 +124,7 @@ public function setDescription($description) {
 	 * @since 14.0.0
 	 */
 	public function setProviderId($providerId) {
-		$this->share['providerId'] = $providerId;
+		$this->share['providerId'] = (string)$providerId;
 	}
 
 	/**
