diff --git a/lib/private/Federation/CloudFederationProviderManager.php b/lib/private/Federation/CloudFederationProviderManager.php
index 81b5d717a56..3f3ef39e373 100644
--- a/lib/private/Federation/CloudFederationProviderManager.php
+++ b/lib/private/Federation/CloudFederationProviderManager.php
@@ -111,7 +111,7 @@ class CloudFederationProviderManager implements ICloudFederationProviderManager
 			} catch (OCMProviderException) {
 				return false;
 			}
-			if ($response->getStatusCode() === Http::STATUS_CREATED) {
+			if ($response->getStatusCode() === Http::STATUS_CREATED || $response->getStatusCode() === Http::STATUS_OK) {
 				$result = json_decode($response->getBody(), true);
 				return (is_array($result)) ? $result : [];
 			}
