diff --git a/lib/private/User/Manager.php b/lib/private/User/Manager.php
index adffe09f2e8..2cbf74b2f50 100644
--- a/lib/private/User/Manager.php
+++ b/lib/private/User/Manager.php
@@ -706,7 +706,7 @@ class Manager extends PublicEmitter implements IUserManager {
                }
 
                // User ID is too long
-               if (strlen($uid) > IUser::MAX_USERID_LENGTH) {
+               if (strlen($uid) > 256) {
                        // TRANSLATORS User ID is too long
                        throw new \InvalidArgumentException($l->t('Username is too long'));
                }
